Add RedundantElseInWhen rule

3flex opened the pull request:
Complements MissingWhenCase rule (#1602)

schalkms commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/RedundantElseInWhen.kt:
```diff
@@ -0,0 +1,103 @@
+        if (expression.entries.find { it.isElse } == null) return
+        if (expression.isUsedAsExpression(bindingContext)) return
+        val subjectExpression = expression.subjectExpression ?: return
+        val subjectType = subjectExpression.kotlinType(bindingContext) ?: return
```
It's hard to understand for me what the code from line 86-88 does and why we need it.
Maybe we can refactor this into a own method and add a descriptive name with a comment?

3flex replied:
I'm going to review this more closely and add additional tests to make sure the rule is only active when required - looking more closely at the core this is adapted from I think there might be a couple of logical issues.

Thanks for challenging this.

schalkms commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/RedundantElseInWhen.kt:
```diff
@@ -0,0 +1,103 @@
+        val subjectExpression = expression.subjectExpression ?: return
+        val subjectType = subjectExpression.kotlinType(bindingContext) ?: return
+
+        if (WhenChecker.getMissingCases(expression, bindingContext).isEmpty() && !subjectType.isFlexible()) {
```
`subjectType.isFlexible()` -> what's meant by that?

schalkms commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/RedundantElseInWhen.kt:
```diff
@@ -0,0 +1,103 @@
+                bindingContext[DECLARATION_TO_DESCRIPTOR, subjectClass?.toSourceElement?.getPsi()]
+            if (subjectClass == null ||
+                KotlinBuiltIns.isBooleanOrNullableBoolean(subjectType) ||
+                subjectClass.module == pseudocodeDescriptor?.module
```
Could you please explain to me the last check in this if expression?

3flex replied:
This is used because code in other modules not in the developer's control might unexpectedly add a new enum value (e.g. when checking cases based on enum or sealed class in an external modules) so it might be desirable to guard against that with an `else`. This was the result of discussion in KT-17497.

detekt can be more flexible here so i will introduce a "strict" config option to either enable or disable that check.

3flex commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/RedundantElseInWhen.kt:
```diff
@@ -0,0 +1,103 @@
+ * </compliant>
+ *
+ * Based on code from Kotlin compiler:
+ * https://github.com/JetBrains/kotlin/blob/v1.3.30/compiler/frontend/src/org/jetbrains/kotlin/cfg/ControlFlowInformationProvider.kt
```
No, I intended to add it - does it need to go before the compliant/noncompliant blocks?

schalkms replied:
Yes, it does.

3flex commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/RedundantElseInWhen.kt:
```diff
@@ -0,0 +1,103 @@
+
+    override val issue: Issue = Issue(
+        "RedundantElseInWhen",
+        Severity.Defect,
```
No, sorry, bad copy paste. Will change to `Warning` unless you think something more appropriate?

Off topic: we should document a definition for each of the Severity enum values.

schalkms replied:
Good.
Thanks for submitting the issue about the docs right away.

3flex commented:
I've addressed some feedback, but unfortunately won't have the time to dedicate to the rest of it. If this isn't suitable to merge as is I'm happy for someone else to pick up my branch and continue from there.

arturbosch commented:
@3flex I was just waiting for the rebase ^^ thanks!
@schalkms why would you wait for 1.0 with this PR?
It's a rule which is only run when type resolution is active which is experimental and needs special flags.

schalkms commented:
> It's a rule which is only run when type resolution is active which is experimental and needs special flags.

You are right. I forgot about this.
LGTM then
